activerecord/test/cases/associations/nested_through_associations_test.rb

```diff
@@ -610,6 +610,12 @@ def test_polymorphic_has_many_through_joined_different_table_twice
     assert_equal hotel, Hotel.joins(:cake_designers, :drink_designers).take
   end
 
+  def test_has_many_through_reset_source_reflection_after_loading_is_complete
+    preloaded = Category.preload(:ordered_post_comments).find(1, 2).last
+    original = Category.find(2)
+    assert_equal original.ordered_post_comments.ids, preloaded.ordered_post_comments.ids
```

Comment:
```suggestion
    assert_equal original.posts.first.comments.ids, preloaded.posts.first.comments.ids
```

Reply:
I don't think this is right assertion to have.

Here I check the following situation:

The has many through association of the second tier (aka "has many through through") may have same records in the through association:

```
category(1).posts.first.comments.first == category(2).posts.first.comments.first
```

Note that this can never happen for through association in the 1st tier.

When we reset the source reflection after processing each owner (e.g. category), the following 
owner may not be able to use the same through record (shared comment) because source reflection would already be reset.

That is why reset code for the source reflection was moved out of `owners.each` loop.

Reply:
The reason that I suggested is the assersion isn't failed without this patch in locally.

Reply:
That is true. I had a version of the patch with only one line of code changed:

``` diff
- association.reset if preload_scope	
+ association.reset if scope
```

However, in a separated branch where I made experiments with other things, I realized this one-line patch adds a regression. That is how this patch end up having 2 tests:

* first test
* one line of code patch
* discovered regression
* second test
* improved patch

I was trying to invent the version of this test that would fail on master but no luck.
You may try it yourself using my description of the problem above. It should theoretically exist.  Maybe with "has many through through through" :).

Reply:
I see. I know it is hard to invent the "has many through through through" failing test.
It is fine as is since this may prevent any regression by any other refactoring in the future.